storage_controller/src/service.rs

```diff
@@ -4012,6 +4048,92 @@ impl Service {
         Ok(())
     }
 
+    async fn activate_timeline_post_import(
+        self: &Arc<Self>,
+        import: &TimelineImport,
+    ) -> anyhow::Result<()> {
+        let mut shards_to_reset: HashSet<ShardIndex> =
+            import.shard_statuses.0.keys().cloned().collect();
+
+        while !shards_to_reset.is_empty() {
+            if self.cancel.is_cancelled() {
+                anyhow::bail!("Shut down requested while finalizing import");
+            }
+
+            let targets = {
+                let locked = self.inner.read().unwrap();
+                let mut targets = Vec::new();
+
+                for (tenant_shard_id, shard) in locked
+                    .tenants
+                    .range(TenantShardId::tenant_range(import.tenant_id))
+                {
+                    if !import
+                        .shard_statuses
+                        .0
+                        .contains_key(&tenant_shard_id.to_index())
+                    {
+                        anyhow::bail!("Shard layout change detected on completion");
```

Comment:
... which should have been prevented through the tenant op lock, right? Important to call out this is an implementation bug

Reply:
> ... which should have been prevented through the tenant op lock, right?

The tenant op lock is not the primary mechanism through which we ensure mutual exclusion of shard splits and imports. See https://github.com/neondatabase/neon/pull/11632.

But, point taken, that's an implementation bug, and, more importantly, a fatal error which should not be retried.
I'll distinguish the errors between retryable and fatal. On fatal errors, we should notify cplane and wait for timeline deletion.

Reply:
Refactored and separated errors in: https://github.com/neondatabase/neon/pull/11855/commits/76a812af39936ad20eb7fe71bc03198654709595